File: src/ast/lexer.cpp
//
// Created by pha on 10/18/23.
//

#include "lexer.h"

bool is_number(const char cur) {
    return cur >= '0' && cur <= '9';
}

void Lexer::consume() {
    this->skip_whitespace();

    if (this->m_position == this->m_source.length()) {
        this->m_token = Token(Type::END, {.none = nullptr });
        return;
    }

    if (is_number(this->peek(0))) {
        this->consume_number();
        return;
    }

    switch (this->peek(0)) {
        case '+':
            this->m_token = Token(Type::PLUS, {.none = nullptr });
            break;
        case '-':
            this->m_token = Token(Type::MINUS, {.none = nullptr });
            break;
        case '*':
            this->m_token = Token(Type::STAR, {.none = nullptr });
            break;
        case '/':
            this->m_token = Token(Type::SLASH, {.none = nullptr });
            break;
        default:
            this->m_token = Token(Type::INVALID, {.none = nullptr });
            break;
    }

    this->m_position++;
}

void Lexer::consume_number() {
    int64_t base = 1;
    int64_t value = 0;

    while (is_number(this->peek(0))) {
        value *= base;
        value  += this->peek(0) - '0';
        base *= 10;

        this->m_position++;
    }

    this->m_token = Token(Type::NUMBER, {.number = value});
}

void Lexer::skip_whitespace() {
    while (this->m_position < this->m_source.length() && this->m_source[this->m_position] == ' ') {
        this->m_position++;
    }
}

char Lexer::peek(size_t offset) {
    return this->m_source[m_position - offset];
}

Token &Lexer::next_token() {
    this->consume();
    return this->m_token;
}

File: src/ast/lexer.rs
use crate::ast::token::Token;

pub struct Lexer {
    source: Vec<char>,
    loc: usize,
}

impl Lexer {
    pub fn new(source: Vec<char>) -> Self {
        Self {
            source,
            loc: 0,
        }
    }

    pub fn next_token(&mut self) -> Option<Token> {
        self.skip_whitespace();

        if self.loc >= self.source.len() {
            return None;
        }

        let token = match self.source[self.loc] {
            '+' => Token::Plus,
            '-' => Token::Minus,
            '*' => if self.peek(1) == '*' {
                self.loc += 1; 
                Token::Power
            } else {
                Token::Star
            },
            '^' => Token::XOR,
            '!' => Token::Not,
            '&' => {
                if self.peek(1) == '&' {
                    self.loc += 1;
                    Token::And
                } else {
                    Token::BitAnd
                }
            }
            '|' => {
                if self.peek(1) == '|' {
                    self.loc += 1;
                    Token::Or
                } else {
                    Token::BitOr
                }
            }
            '/' => Token::Slash,
            '(' => Token::LParen,
            ')' => Token::RParen,
            '=' => {
                if self.peek(1) == '=' {
                    self.loc += 1;
                    Token::Equal
                } else {
                    Token::Assign
                }
            }
            ':' if self.peek(1) == '=' => { self.loc += 1; Token::Decl }
            'A'..='Z' | 'a'..='z' => self.consume_ident(),
            '0'..='9' => self.consume_number(),
            _ => Token::Invalid,
        };

        self.loc += 1;
        Some(token)
    }

    fn peek(&self, offset: usize) -> char {
        if offset + self.loc > self.source.len() {
            println!("hhshh");
            return ' ';
        }

        self.source[self.loc + offset]
    }

    fn skip_whitespace(&mut self) {
        while self.loc < self.source.len() && self.source[self.loc] == ' ' {
            self.loc += 1;
        }
    }

    fn consume_ident(&mut self) -> Token {
        let mut acc = String::new();

        while self.loc < self.source.len() && self.source[self.loc].is_alphanumeric() {
            acc.push(self.source[self.loc]);
            self.loc += 1;
        }

        match acc.as_str() {
            "true" | "false" => Token::Bool(acc.as_str() == "true"),
            _ => Token::Ident(acc),
        }
    }

    fn consume_number(&mut self) -> Token {
        let src = self.loc;

        while self.loc + 1 < self.source.len() && self.source[self.loc + 1].is_numeric() {
            self.loc += 1;
        }

        let num = self.source[src..self.loc + 1].iter().collect::<String>();

        Token::Number(num.parse::<i64>().unwrap())
    }
}
